### ISSUE TITLE
"testname" ant attribute doesn't work

### ISSUE THREAD

**Issue by anupchandak** [cbeust/testng#800]
I am trying to invoke tests of a testng.xml file from ant. 
I want to invoke it for particular 'test' so I used testname="myTestName"
But this is not working. It by default runs all the test cases specified in tesng.xml under all 'test' tag

**Comment by juherr** [comment]
Could you share a sample project which can be executable?

**Comment by anupchandak** [comment]
Hi - I have created a simple project having two test classes TestNGTest1 & TestNGTest1. Each of them are simply printing SoP. 
Each of these classes are called under different ***test*** tag in testng.ml
This project is packaged as jar file and being executed via build file external to it. 

Whenever I run target *run-unit-test* from my build.xml, I get SoP from both classes irrespective of what value I set for ***testnames*** attribute.

**Comment by juherr** [comment]
Could you share this project? It will really help ;)

**Comment by anupchandak** [comment]
Yeah. First time I am using GitHub to do sharing so setting up repository. Give me some time.

**Comment by anupchandak** [comment]
I have created public repository (https://github.com/anupchandak/Test) where you can find a build.xml and all other required lib for reproducing above issue.
If you want to look into testNG code whichi have written for this test thenyou can check at (https://github.com/anupchandak/TestNGTest)

I hope this helps you. IF you need any more information then do let me know.

One more thing. I am able to use this attribute when i do execution in Eclipse workbench. It fails when I package my project and I try to run a JAR having testNG classes. 

Thanks!

**Comment by juherr** [comment]
Ok, thank.

First thing, could you try with the latest TestNG version (6.9.6)?

**Comment by juherr** [comment]
`testnames` is not doing what you expect. [The documentation](http://testng.org/doc/ant.html) says: 
> A comma separated list of test names, as defined in the <test> tag. Only these tests will be run. Defaults to "Ant test"

What you should use is `classfilesetref`.

**Comment by anupchandak** [comment]
Hello! Ok. I will surely try to see if I can use `<classfilesetref>`. Do you have any handy example about its usage?

But still I have a question is why then behavior of `<testnames>` is different when it is used in Eclipse workbench than when its used for packaged JAR? Isn't that should be same? 

Moreover, I do not think that I have used `<testnames>` wrongly. As documentation says I can provide comma separated list. But in my case if I want to invoke just one then that list will have only one entry (so indirectly no comma). Correct me, if I am wrong.

**Comment by juherr** [comment]
My bad, you right. It was a bit too late yesterday. 

BTW, you can try to use the long name of classes. 
For examples, you can ask on stackoverflow.com or on the mailing list. I haven't used ant for years.

**Comment by anupchandak** [comment]
HI - Yes, I got the way to use `<classfilesetref>` but frankly speaking, ideally I would not like to use that. Mine is a complex application with around 25 to 30 different projects with number of people working on it across glob and adding test cases. For simplicity we planned to categorized them in two sets - Unit Test and Integration Test. If I start keeping track of classes then it will be too much of work for Build Team. 

Let me know if you will be opening this issue and if any fixing will be done to this issue. 

Meanwhile, I will explore if I can create some reg ex and use `<classfilesetref>` without burdening build team to know each time developer adds a new test classes. Thanks!

**Comment by juherr** [comment]
In fact, `<testnames>` with ant should work [like the command line](http://testng.org/doc/documentation-main.html#running-testng) because it is what the ant taks in doing.

According the code, `<testnames>` are the provided names of `<test>` from the `testng.xml`.

**Comment by anupchandak** [comment]
Yes, so will you be taking this as defect? and opening it ?

**Comment by anupchandak** [comment]
Moreover I as well tried using `<classfileset >`, but no luck.

	<target name="run-unit-test">
		<testng classpathref="Test.classpath" testjar="TestNGTest.jar">
			<classfileset dir=".">
				<include name="**/*1.class"/>
			</classfileset>
		</testng>
	</target>

**Comment by juherr** [comment]
As I understand, `testnames` need a `testng.xml`.

**Comment by anupchandak** [comment]
yes- `testng.xml` is very much part of the JAR (which I shared). kept at its root.

**Comment by juherr** [comment]
Ok, I think I understand the use case.
I found a test which should test the feature (`JarPackagesTest`) which is never run.

The test has to be added in the suite, fixed and we will see if it reproduce the issue.
I can't tell when I will do it. But if you want try, you're welcome :)

**Comment by juherr** [comment]
This looks like a usage question rather than a TestNG bug, and it has been inactive for a long time. For usage help please use the [TestNG user group](https://groups.google.com/forum/#!forum/testng-users) or [StackOverflow](https://stackoverflow.com/questions/tagged/testng). Closing as stale — feel free to reopen with a minimal reproducible example on the latest TestNG version.
